### PR TITLE
feat(storage-browser): scaffold component

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
Scaffold `StorageBrowser`:
- add `createStorageBrowser`
- add `listLocations` and `listLocationItems` actions and types
- add `LocationsView` and `LocationView` with mock data fetching
- add iniital `element` definitions
- add mock data generation utils
- add NextJS example pages for standalone composability testing, `LocationsView` and `LocationView`
- add `PaginateState` and `HistoryState` context
- expose React utils from ui-react-core, ui-react
- add basic unit tests
- configure CI and feature branch tag publishing\

#### Issue #, if available
NA

#### Description of how you validated changes
- manual testing
- `yarn test`
#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added
- [x] PR title and commit messages follow [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) syntax

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-storage/src/components/StorageBrowser/context/actions/listLocationItemsAction.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    refresh?: boolean;
+  };
+}
+
+interface ListLocationItemsActionOutput {
+  items: LocationItem[];
+  nextToken: string | undefined;
+}
+
+const generateListLocationItemsData = (count: number): LocationItem[] => {
```
Comment (timngyn) [inline packages/react-storage/src/components/StorageBrowser/context/actions/listLocationItemsAction.ts:30]: just for my understanding, is this function just a way to help you generate location data?

Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/context/actions/listLocationItemsAction.ts:30]: Yes, added it for development just for rendering data. Will be used for unit tests in the future

File: packages/react-storage/src/components/StorageBrowser/context/primitives.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  DefaultPrimitives,
+} from '@aws-amplify/ui-react/internal';
+
+export interface StorageBrowserPrimitives
+  extends Pick<DefaultPrimitives, 'Button' | 'View'> {}
+
+const defaultValue: StorageBrowserPrimitives = {
+  Button: PrimitiveButton,
+  View: PrimitiveView,
+};
```
Comment (timngyn) [inline packages/react-storage/src/components/StorageBrowser/context/primitives.ts:14]: sorry if this was mentioned somewhere else, but what are the Button and View supposed to represent? This context isn't being used in this PR at the moment right?

**Review by timngyn — APPROVED** [review]
Looks good to me! Just had one question for my understanding

File: packages/react-storage/src/components/StorageBrowser/context/primitives.ts (hunk 2 above)
Comment (calebpollman) [reply packages/react-storage/src/components/StorageBrowser/context/primitives.ts:14]: That's correct, this is just scaffolding/adding as examples

**Review by reesscot — DISMISSED** [review]

File: packages/react-storage/src/components/StorageBrowser/context/actions/listLocationsAction.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  // bucket: <bucket>/*
+  const bucketName = `${generateString(randomNumberInRange(10, 24))}/`;
+  const hasPrefixes = !!((randomNumberInRange(0, 31) / 2) % 2);
+  const prefixCount = !hasPrefixes ? 0 : randomNumberInRange(1, 4);
+
+  const prefixes = Array(prefixCount)
+    .fill('')
+    .map(
+      // prefix: <bucket>/<prefix-with-path>*
+      () => `${bucketName}${generateString(randomNumberInRange(6, 20))}`
```
Comment (AllanZhengYP) [inline packages/react-storage/src/components/StorageBrowser/context/actions/listLocationsAction.ts:41]: missing ending `*`?

File: packages/react-storage/src/components/StorageBrowser/context/elements/definitions.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  displayName: 'Text',
+});
+
+export const OrderedListElement = defineBaseElement({
+  type: 'ol',
+  displayName: 'OrderedList',
+});
+
+export const UnorderedListElement = defineBaseElement({
+  type: 'ol',
```
Comment (hbuchel) [inline packages/react-storage/src/components/StorageBrowser/context/elements/definitions.ts:71]:
```suggestion
  type: 'ul',
```

**Review by hbuchel — DISMISSED** [review]

**Review by hbuchel — APPROVED** [review]

**Review by timngyn — APPROVED** [review]

**Review by thaddmt — COMMENTED** [review]
Had some small questions, can follow up offline!

**Pull request merged**